src/tests/roc_fec/target_openfec/test_writer_reader.cpp

```diff
@@ -493,44 +599,43 @@ TEST(writer_reader, decoding_when_multiple_blocks_in_queue) {
 
         PacketDispatcher dispatcher;
 
-        Writer writer(config, FECPayloadSize, encoder, dispatcher, source_composer,
-                      repair_composer, packet_pool, buffer_pool, allocator);
+        packet::Interleaver intrlvr(dispatcher, allocator, 10);
+
+        CHECK(intrlvr.valid());
+
+        Writer writer(config, FECPayloadSize, encoder, intrlvr, source_composer(),
+                      repair_composer(), packet_pool, buffer_pool, allocator);
 
         Reader reader(config, decoder, dispatcher.source_reader(),
                       dispatcher.repair_reader(), rtp_parser, packet_pool, allocator);
 
         CHECK(writer.valid());
         CHECK(reader.valid());
 
-        for (size_t block_num = 0; block_num < NumBlocks; ++block_num) {
-            fill_all_packets(NumSourcePackets * block_num);
+        packet::PacketPtr many_packets[NumPackets];
 
-            for (size_t i = 0; i < NumSourcePackets; ++i) {
-                writer.write(source_packets[i]);
-            }
+        for (size_t i = 0; i < NumPackets; ++i) {
+            many_packets[i] = fill_one_packet(i);
+            writer.write(many_packets[i]);
         }
-        dispatcher.push_all();
-
-        CHECK(dispatcher.source_size() == NumSourcePackets * NumBlocks);
-        CHECK(dispatcher.repair_size() == NumRepairPackets * NumBlocks);
+        dispatcher.push_written();
 
-        for (size_t block_num = 0; block_num < NumBlocks; ++block_num) {
-            for (size_t i = 0; i < NumSourcePackets; ++i) {
-                packet::PacketPtr p = reader.read();
-                CHECK(p);
-                check_audio_packet(p, NumSourcePackets * block_num + i);
-            }
+        intrlvr.flush();
 
-            dispatcher.reset();
+        for (size_t i = 0; i < NumPackets; ++i) {
+            packet::PacketPtr p = reader.read();
+            CHECK(p);
+            check_audio_packet(p, i);
+            check_restored(p, false);
         }
     }
 }
 
-IGNORE_TEST(writer_reader, decoding_late_packet) {
-    // 1. Fill all packets in block except one lost packet.
-    // 2. Read first part of block till lost packet.
-    // 3. Receive one missing packet.
-    // 4. Read and check latter block part.
+TEST(writer_reader, delayed_packets) {
```

Comment:
Because this test models the case when a bunch of packets is delayed by network.

We first deliver and read first 10 packets.

Then try to read the 11th packet and get NULL.

Then we deliver the rest packets and ensure that we will read all of them, starting from 11th.

So basically this test checks that if there was no packet when it's time to read it, and also there were no packets after it, the unsuccessful read will not advance reader position and when the packets are delivered we will be able to read them.

BTW this case was added because it needs special handling in reader.

I'll add a comment to the test.

Reply:
> I'll add a comment to the test.

Done.